# Solution for Homework 1

## Problem 1

The file "Temp.txt" which you can download from the homework tab on Canvas, contains the monthly high-temperatures (in degrees Celsius) at Heathrow airport for the years 1948 through 2016. Save the file on your computer (in a folder that you can find again). Run the code below, changing the path to the appropriate path on **your** computer. It's ok to use forward slashes in the path name, regardless of whether you're working on Windows or MacOS. 

In [1]:
temperatures = []

from pathlib import Path

data_folder = Path('/Users/mbremer/SJSU/Math167PS/Data/HomeworkData')
# make sure to change this path to the actual folder where YOU have saved the data file. 

file_to_open = data_folder / "Temp.txt"

file_to_open
f = open(file_to_open)

for row in f:
    temperatures.append(float(row))

You have now created an object called temperatures. 

Answer the following questions by inserting your code into the code box (after the "# your code here" prompt. Comment out or delete the "raise NotImplementedError()") and answering the question in the form of a complete English sentence in the respective Markdown box. 

a) What kind of object type is "temperatures"?

In [2]:
type(temperatures)

list

`temperatures` is a list object. 

b) Describe what the "float()" and "append()" commands do, respectively, that have been applied above. Try to leave out the "float()" command. What happens then? 

The "float()" method translates the entries in the rows of the file temperature into float objects (that is, numbers). Without it, the data would be read in as strings complete with line-break characters. The append command creates a list that adds the recorded temperatures row by row. That is, each row will become one entry. 

c) How many temperatures have been recorded in the file?

In [3]:
len(temperatures)

828

828 temperatures have been recorded. 

d) What are the highest and lowest temperatures recorded?

In [4]:
print(min(temperatures))
print(max(temperatures))

0.8
28.2


The lowest recorded monthly high temperature is 0.8 degrees Celsius and the highest recorded monthly high temperature is 28.2 degrees Celsius in this file. Instead of using min and max functions, you could have also sorted the list and looked at the first and last entry, respectively. 

e) The first observation in the `temperatures` object is from a January 1948. Using only the commands you learned from chapter 5 (that is, using base python and no additional python libraries), find the the mean January high temperature at London Heathrow between the years 1948 and 2016. Have your code round your answer to two decimal digits accuracy and return the answer as a float.  

In [5]:
round(sum(temperatures[::12])/len(temperatures[::12]),2)

7.56

f) Find the median monthly high temperature (using only the commands you learned from chapter 5 and base python, not using additional modules or libraries). Use Python to round your result to two after-decimal digits. State the result and explain briefly how you find the median in the markdown box below. 

In [6]:
temperatures.sort()
temperatures[413:415]

[14.7, 14.7]

Since the number of observations is even, the median is the average of the two "middle" observations which are both 14.7. Hence, the median monthly high temperature is 14.7 degrees Celsius. 

## Problem 2 

Much of data science these days is based on text data. The text to be analyzed may be literature or postings on social media. The tasks to be performed could be sentiment analysis (trying to guess whether the writing is "pro" or "con" on a certain issue) to automatically detecting postings by Malicious bots. However, before you can perform any kind of machine learning procedure on a text, you have to first learn how to "clean" the text. That usually means removing all non-text elements and sometimes also removing filler words such as "the", "it", "and" etc. After that, what is usually analyzed is the frequency with which certain words appear and most often actually not the *order* in which they appear. 

The text below are the first three paragraphs of Franz Kafka's book "Metamorphosis". Our goal is to clean this text passage. 

In [8]:
text = """One morning, when Gregor Samsa woke from troubled dreams, he found
himself transformed in his bed into a horrible vermin.  He lay on
his armour-like back, and if he lifted his head a little he could
see his brown belly, slightly domed and divided by arches into stiff
sections.  The bedding was hardly able to cover it and seemed ready
to slide off any moment.  His many legs, pitifully thin compared
with the size of the rest of him, waved about helplessly as he
looked.

"What's happened to me?" he thought.  It wasn't a dream.  His room,
a proper human room although a little too small, lay peacefully
between its four familiar walls.  A collection of textile samples
lay spread out on the table - Samsa was a travelling salesman - and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice, gilded frame.  It showed
a lady fitted out with a fur hat and fur boa who sat upright,
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer.

Gregor then turned to look out the window at the dull weather.
Drops of rain could be heard hitting the pane, which made him feel
quite sad.  "How about if I sleep a little bit longer and forget all
this nonsense", he thought, but that was something he was unable to
do because he was used to sleeping on his right, and in his present
state couldn't get into that position.  However hard he threw
himself onto his right, he always rolled back to where he was.  He
must have tried it a hundred times, shut his eyes so that he
wouldn't have to look at the floundering legs, and only stopped when
he began to feel a mild, dull pain there that he had never felt
before."""

a) Look at the text. What non-alphabetic character elements do you see in the text?

The non-character elements in this text (that you can see) are: . , " ' - ? There are also "\\n" and "\\t" characters (which you cannot see). 

b) Using only the commands covered in class so far (in base python without use of additional libraries), write code to remove all non-alphabetic characters (such as punctuation or hyphens).

In [9]:
text = text.replace(".","")
text = text.replace(",","")
text = text.replace("\'","")
text = text.replace('"',"")
text = text.replace('?',"")
text = text.replace("-"," ")
#text = text.replace("\n", " ") # optional, the split command will also take care of these. 
#text = text.replace("\t", "")

c) Write code to make everything lower case and convert the text into a list of single words. 

In [10]:
text = text.lower()
word_list = text.split()

d) How many total words are in the list (if armour-like is considered to be two words)? Count, how often the word "he" appears in the three paragraphs. Don't count if "he" is part of another word, e.g., "tHEn". 

In [11]:
print(len(word_list))
word_list.count("he")

314


17

There are 314 words. The word "he" appears 17 times. 

e) How many distinct (different) words appear in the first three paragragraphs?

In [12]:
unique_words = set(word_list)
len(unique_words)

195

There are 195 unique words. 